File: app/src/main/java/com/cv4j/app/activity/HistogramEqualizationActivity.java
package com.cv4j.app.activity;

import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.widget.ImageView;

import com.cv4j.app.R;
import com.cv4j.core.datamodel.ByteProcessor;
import com.cv4j.core.datamodel.CV4JImage;
import com.cv4j.core.datamodel.ImageProcessor;
import com.cv4j.core.hist.CalcHistogram;
import com.cv4j.core.hist.EqualHist;
import com.safframework.injectview.annotations.InjectExtra;
import com.safframework.injectview.annotations.InjectView;
import com.safframework.injectview.annotations.OnClick;

/**
 * Created by Tony Shen on 2017/5/14.
 */

public class HistogramEqualizationActivity extends BaseActivity {

    @InjectView(R.id.image0)
    ImageView image0;

    @InjectView(R.id.calc_image0)
    ImageView calcImage0;

    @InjectView(R.id.image1)
    ImageView image1;

    @InjectView(R.id.calc_image1)
    ImageView calcImage1;

    @InjectView(R.id.toolbar)
    Toolbar toolbar;

    @InjectExtra(key = "Title")
    String title;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_histogram_equalization);

        initData();
    }

    private void initData() {

        toolbar.setTitle("< "+title);
        Resources res = getResources();
        final Bitmap bitmap = BitmapFactory.decodeResource(res, R.drawable.test_hist);
        image0.setImageBitmap(bitmap);

        CV4JImage cv4jImage0 = new CV4JImage(bitmap);
        ImageProcessor imageProcessor = cv4jImage0.convert2Gray().getProcessor();
        Paint paint = new Paint();
        calcImage0.setImageBitmap(drawHist(imageProcessor,paint));

        CV4JImage cv4jImage = new CV4JImage(bitmap);
        imageProcessor = cv4jImage.convert2Gray().getProcessor();
        if (imageProcessor instanceof ByteProcessor) {
            EqualHist equalHist = new EqualHist();
            equalHist.equalize((ByteProcessor) imageProcessor);
            image1.setImageBitmap(cv4jImage.getProcessor().getImage().toBitmap());
            paint = new Paint();
            calcImage1.setImageBitmap(drawHist(imageProcessor,paint));
        }
    }

    private Bitmap drawHist(ImageProcessor imageProcessor,Paint paint) {

        CalcHistogram calcHistogram = new CalcHistogram();
        int bins = 127;
        int[][] hist = new int[imageProcessor.getChannels()][bins];
        calcHistogram.calcRGBHist(imageProcessor,bins,hist,true);
        Bitmap bm = Bitmap.createBitmap(512,512, Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bm);
        paint.setColor(Color.BLACK);
        paint.setStyle(Paint.Style.FILL_AND_STROKE);
        canvas.drawRect(0,0,512,512,paint);

        float step = 512.0f/127;
        int xoffset;
        int yoffset;
        int channels = imageProcessor.getChannels();

        int[] colors = new int[]{Color.argb(127,255,0,0),Color.argb(127,0,255,0),Color.argb(127,0,0,255)};
        for (int i=0;i<channels;i++) {

            paint.setColor(colors[i]);
            for (int j=0;j<bins;j++) {

                xoffset = (int)(j*step);
                yoffset = hist[i][j]*512/255;
                canvas.drawRect(xoffset,512-yoffset,xoffset+step,512,paint);
            }
        }

        return bm;
    }

    @OnClick(id= R.id.toolbar)
    void clickToolbar() {

        finish();
    }
}

File: app/src/main/java/com/cv4j/app/activity/HistogramEqualizationActivity.kt
package com.cv4j.app.activity

import android.content.res.Resources
import android.graphics.*
import android.os.Bundle
import com.cv4j.app.R
import com.cv4j.app.app.BaseActivity
import com.cv4j.core.datamodel.ByteProcessor
import com.cv4j.core.datamodel.CV4JImage
import com.cv4j.core.datamodel.ImageProcessor
import com.cv4j.core.hist.CalcHistogram
import com.cv4j.core.hist.EqualHist
import kotlinx.android.synthetic.main.activity_histogram_equalization.*

/**
 *
 * @FileName:
 *          com.cv4j.app.activity.HistogramEqualizationActivity
 * @author: Tony Shen
 * @date: 2020-05-05 16:28
 * @version: V1.0 <描述当前版本功能>
 */

class HistogramEqualizationActivity : BaseActivity() {

    var title: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_histogram_equalization)

        intent.extras?.let {
            title = it.getString("Title","")
        }?:{
            title = ""
        }()

        toolbar.setOnClickListener {
            finish()
        }

        initData()
    }

    private fun initData() {
        toolbar.title = "< $title"
        val res: Resources = resources
        val bitmap = BitmapFactory.decodeResource(res, R.drawable.test_hist)
        image0.setImageBitmap(bitmap)
        val cv4jImage0 = CV4JImage(bitmap)
        var imageProcessor = cv4jImage0.convert2Gray().processor
        var paint = Paint()
        calc_image0.setImageBitmap(drawHist(imageProcessor, paint))
        val cv4jImage = CV4JImage(bitmap)
        imageProcessor = cv4jImage.convert2Gray().processor
        if (imageProcessor is ByteProcessor) {
            val equalHist = EqualHist()
            equalHist.equalize(imageProcessor)
            image1.setImageBitmap(cv4jImage.processor.image.toBitmap())
            paint = Paint()
            calc_image1.setImageBitmap(drawHist(imageProcessor, paint))
        }
    }

    private fun drawHist(imageProcessor: ImageProcessor, paint: Paint): Bitmap {
        val calcHistogram = CalcHistogram()
        val bins = 127
        val hist = Array(imageProcessor.channels) { IntArray(bins) }
        calcHistogram.calcRGBHist(imageProcessor, bins, hist, true)
        val bm = Bitmap.createBitmap(512, 512, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bm)
        paint.color = Color.BLACK
        paint.style = Paint.Style.FILL_AND_STROKE
        canvas.drawRect(0f, 0f, 512f, 512f, paint)
        val step = 512.0f / 127
        var xoffset: Int
        var yoffset: Int
        val channels = imageProcessor.channels
        val colors = intArrayOf(Color.argb(127, 255, 0, 0), Color.argb(127, 0, 255, 0), Color.argb(127, 0, 0, 255))
        for (i in 0 until channels) {
            paint.color = colors[i]
            for (j in 0 until bins) {
                xoffset = (j * step).toInt()
                yoffset = hist[i][j] * 512 / 255
                canvas.drawRect(xoffset.toFloat(), 512 - yoffset.toFloat(), xoffset + step, 512f, paint)
            }
        }
        return bm
    }
}